fastjsonschema/generator.py

```diff
@@ -169,10 +181,30 @@ def generate_func_code(self, definition):
         Creates base code of validation function and calls helper
         for creating code by definition.
         """
-        with self.l('def func(data):'):
-            self.l('NoneType = type(None)')
+        self._validation_functions_done.add(self._resolver.get_uri())
+        self.l('NoneType = type(None)')
+        self.l('')
+        with self.l('def validate(data):'):
             self.generate_func_code_block(definition, 'data', 'data')
             self.l('return data')
+        # Generate parts that are referenced and not yet generated
+        while len(self._needed_validation_functions) > 0:
+            # Normal for ... in... loop not works here, because new functions
+            # can be found and therefore added in self._needed_validation_functions
+            # during it generation and dictionary cannot changed during iteration.
+            uri, name = self._needed_validation_functions.popitem()
+            self._validation_functions_done.add(uri)
+            self.generate_validation_function(uri, name)
+
+    def generate_validation_function(self, uri, name):
+        """
+        Generate validation function for given uri with given name
+        """
+        self.l('')
+        with self._resolver.resolving(uri) as definition:
+            with self.l('def {}(data):', name):
+                self.generate_func_code_block(definition, 'data', 'data', clear_variables=True)
+                self.l('return data')
```

Comment:
I'm thinking we could move `self._validation_functions_done.add(uri)` to this function and generate the main function by this one as well. What do you think? :-)

Reply:
Sounds good... I'll make some refactoring :-)

Reply:
It seems that, if main schema has `id` in main level it generates same function with 2 names... It is also refactor away in commit 51b5229.

Reply:
Very nice. It's very clear now.